File: TwoPointer/393.utf-8-validation.go
package leetcode

// @leet start
func validUtf8(data []int) bool {
	var remain int

	for _, b := range data {
		c, mask := 0, 128

		for b&mask != 0 {
			c++
			if c > 4 {
				return false
			}
			mask = mask >> 1

		}

		if c == 0 {
			if remain != 0 {
				return false
			}
			continue
		}

		if c > 1 {
			if remain != 0 {
				return false
			}
			remain = c - 1
			continue
		}

		if remain == 0 {
			return false
		}

		remain--

	}

	return remain == 0
}

// @leet end

File: TwoPointer/393.utf-8-validation.py
from typing import List


# @leet start
class Solution:
    def validUtf8(self, data: List[int]) -> bool:
        remain = 0
        for byte in data:
            c = 0
            mask = 128  # 0b1000000

            while byte & mask != 0:
                c += 1
                mask = mask >> 1
                if c > 4:
                    return False

            if c != 1:
                if remain != 0:
                    return False
                if c > 1:
                    remain = c - 1
                continue

            if remain == 0:
                return False

            remain -= 1

        return remain == 0


# @leet end

